src/net_processing.cpp

```diff
@@ -760,6 +764,12 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
     bool fWitnessEnabled = IsWitnessEnabled(pindex->pprev, Params().GetConsensus());
     uint256 hashBlock(pblock->GetHash());
 
+    {
```

Comment:
Make this a member of PeerLogicValidation, then you can use a weak_ptr for the global, and I think no locks are needed